File: 815 Bus Routes/Solution.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target) return 0;
        Map<Integer, List<Integer>> map = new HashMap<>();  // Key:stop, Value:list of bus travels this stop
        
        for (int i = 0; i < routes.length; i++) {
            for (int j : routes[i]) {
                map.putIfAbsent(j, new ArrayList<Integer>());
                map.get(j).add(i);
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        Set<Integer> visited = new HashSet<>(); // Store bus

        int steps = 0;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int cur = queue.poll();
                if (cur == target) return steps;
                for (int bus : map.get(cur)) {
                    if (visited.contains(bus)) continue;
                    visited.add(bus);
                    for (int next : routes[bus]) {
                        if (next != cur) queue.add(next);
                    }
                }
            }
            steps++;
            
        }
        
        return -1;
    }

}

File: 815 Bus Routes/Solution.cpp
class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int source, int target) {
        if (source == target) return 0;
        
        unordered_map<int, vector<int>> map;
        for (int i = 0; i < routes.size(); ++i)
            for (int stop : routes[i])
                map[stop].push_back(i);
        
        set<int> visited;
        queue<int> q;
        q.push(source);
        int steps = 0;
        
        while (!q.empty()) {
            int size = q.size();
            
            while (size-- > 0) {
                int cur = std::move(q.front());
                if (cur == target) return steps;
                q.pop();
                for (int bus : map[cur]) {
                    if (!visited.count(bus)) {
                        visited.insert(bus);
                        for (int stop : routes[bus]) {
                            q.push(stop);
                        }
                    }
                }
            }
            ++steps;
        }
        return -1;
    }
};